sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -246,6 +247,16 @@ object ColumnPruning extends Rule[LogicalPlan] {
 
       Project(projectList, Join(pruneJoinChild(left), pruneJoinChild(right), joinType, condition))
 
+    // Eliminate unneeded attributes from either side of a Join.
+    case Project(projectList, Filter(predicates, Join(left, right, joinType, condition)))  =>
```

Comment:
Is it a problem for all operator under `Filter`? like `Project<-Filter<-Join`, `Project<-Filter<-Aggregate`, `Project<-Filter<-Sort`, etc.